exploration/code-mode-introducer.md

```diff
@@ -0,0 +1,224 @@
+I want the syntax to be logical and as consistent as possible.
+
+As a translator, I don't want to have to learn special syntax to support features such as declarations.
+
+As a user, I want my messages to be robust.
+Minor edits and changes should not result in syntax errors.
+
+As a user, I want to be able to see which messages are complex at a glance
+and to parse messages into their component parts visually as easily as possible.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+Some of the options use a new sigil as part of the introducer.
+For various reasons, `#` has been used recently as a placeholder for this sigil.
+There are concerns that this character is not suitable, since it is used as a comment
+introducer in a number of formats. 
+See for example [#520](https://github.com/unicode-org/message-format-wg/issues/520).
+The actual sigil used needs to be an ASCII character in the reserved or private use
+set (with syntax adjustments if we use up a private-use one).
+Most of the options below have been changed to use `^`, using 
+Apple's experimental syntax as a model for sigil choice.
+
+It should be noted that an introducer sigil should be as rare as possible in normal text.
+This tends to run against common punctuation marks `&`, `%`, `!`, and `?`.
+
+```abnf
+reserved-start = "!" / "@" / "#" / "%" / "*" / "<" / ">" / "/" / "?" / "~"
+private-start  = "^" / "&"
+```
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+We need to choose one of these (or another option not yet considered).
+Presentation at UTW did not produce any opinions.
+
+Based on the pro/cons below, I would suggest Option D is possibly the best option?
+
+## Alternatives Considered
+
+_What other solutions are available?_
+_How do they compare against the requirements?_
+_What other properties they have?_
+
+There are the following designs being considered:
+
+### Option A. Use Pattern Quotes for Messages (current design)
+
+Complex messages are quoted with double curly brackets.
+The closing curly brackets might be optional.
+
+Sample pattern:
+```
+{{
+input {$var}
+match {$var}
+when * {{Pattern}}
+}}
+```
+Sample quoted pattern with no declarations or match:
+```
+{{{{Pattern}}}}
+```
+
+Pros:
+- Uses a sigil `{` already present in the syntax
+- No additional escapes
+- Consistent with other parts of the syntax?
+
+Cons:
+- Somewhat verbose
+- Closing portion of the syntax adds no value;
+  could be a source of unintentional syntax errors
+- Messages commonly end with four `}}}}`
+
+### Option B. Use a Sigil
+
+Complex messages start with a special sigil character.
+
+```
+^input {$var}
+match {$var}
+when * {{Pattern}}
+```
+Sample quoted pattern with no declarations or match:
+```
+^{{Pattern}}
```

Comment:
In all options not using `{{` for code-mode, we could allow `{{Pattern}}` to be a valid shorthand for quoted patterns.

Reply:
We could, but then we'd be into having multiple equivalent representations?

Reply:
Well, we already have two anyways: `Pattern` and `{{{{Pattern}}}}`. I'm merely describing options that we have. We could, for instance, also consider forbidding complex message syntax without declarations, so that only `Pattern` is valid.

I think the question to ask first is: given that `Pattern` is valid syntax and that we preserve whitespace, do we expect use-cases for quoting simple patterns?

It would be another way of ensuring leading and trailing whitespace is preserved, using only the MF2 syntax, regardless of what the host format offers. Consider:

```xml
<string xml:space="preserve">"   Hello   "</string>
```
```xml
<string>{{{{   Hello   }}}}</string>
```

Reply:
Depending on how you look at this, it would either mean supporting for a pattern within a pattern, or adding a third syntax under the trenchcoat. I'd really rather not. Either we leave the external spaces for the surrounding syntax, or we don't. Let's not do both if we don't need to.